tests/Doctrine/Tests/CodingStandard/TestCase.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\CodingStandard;
+
+use PHP_CodeSniffer\Config;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Files\LocalFile;
+use PHP_CodeSniffer\Runner;
+use PHPUnit\Framework\TestCase as BaseTestCase;
+use RuntimeException;
+use function array_key_exists;
+use function count;
+use function implode;
+use function sprintf;
+use function strpos;
+use function strrpos;
+use function substr;
+
+abstract class TestCase extends BaseTestCase
+{
+    final public function testValidator() : void
+    {
+        $codeSniffer = $this->createCodeSniffer();
+        $file        = $this->createFile($codeSniffer);
+
+        $this->processFile($file);
+
+        $expectedErrors = $this->getExpectedErrors();
+
+        self::assertSame(
+            count($expectedErrors),
+            $file->getErrorCount(),
+            sprintf(
+                'Expected %d error%s but found %d.',
+                count($expectedErrors),
+                count($expectedErrors) === 1 ? '' : 's',
+                $file->getErrorCount()
+            )
+        );
+
+        foreach ($expectedErrors as $expectedError) {
+            [$expectedErrorName, $expectedErrorLine] = $expectedError;
+
+            self::assertTrue(
+                self::hasError($file, ...$expectedError),
+                sprintf(
+                    'Expected error %s at line %d but it was not found.',
+                    $expectedErrorName,
+                    $expectedErrorLine
+                )
+            );
+        }
+    }
+
+    /**
+     * @return (string|int)[]
+     */
+    abstract protected function getExpectedErrors() : array;
+
+    final public function testFixer() : void
+    {
+        $codeSniffer = $this->createCodeSniffer();
+        $file        = $this->createFile($codeSniffer);
+
+        $this->fixFile($file);
+
+        self::assertStringEqualsFile($this->getSubjectFixedFilename(), $file->fixer->getContents());
+    }
+
+    protected function getSubjectName() : string
+    {
+        // Doctrine\Tests\CodingStandard\SomeSubjectTest
+        return substr(static::class, strrpos(static::class, '\\') + 1, -4);
+    }
+
+    final protected function getSubjectFilename() : string
```

Comment:
You don't seem to use `final protected` methods right now in other classes. Why not making them private?

Reply:
I guess is because it can be used in children classes, but not overwritten.